src/stacked_borrows/diagnostics.rs

```diff
@@ -0,0 +1,338 @@
+use rustc_target::abi::Size;
+
+use crate::helpers::HexRange;
+use crate::stacked_borrows::{err_sb_ub, AccessKind, GlobalStateInner, Permission};
+use crate::Item;
+use crate::SbTag;
+use crate::Stack;
+use crate::ThreadManager;
+
+use rustc_middle::mir::interpret::InterpError;
+
+#[derive(Debug, Default)]
+pub struct AllocHistory {
+    // The time tags can be compressed down to one bit per event, by just storing a Vec<u8>
+    // where each bit is set to indicate if the event was a creation or a retag
+    current_time: usize,
+    creations: smallvec::SmallVec<[Event; 2]>,
+    invalidations: smallvec::SmallVec<[Event; 1]>,
+    protectors: smallvec::SmallVec<[Protection; 1]>,
+}
+
+#[derive(Debug)]
+struct Protection {
+    orig_tag: SbTag,
+    tag: SbTag,
+    span: Span,
+}
+
+#[derive(Debug)]
+struct Event {
+    time: usize,
+    parent: Option<SbTag>,
+    tag: SbTag,
+    range: AllocRange,
+    span: Span,
+}
+
+pub enum TagHistory {
+    Tagged {
+        tag: SbTag,
+        created: (AllocRange, SpanData),
+        invalidated: Option<(AllocRange, SpanData)>,
+        protected: Option<(SbTag, SpanData, SpanData)>,
+    },
+    Untagged {
+        recently_created: Option<(AllocRange, SpanData)>,
+        recently_invalidated: Option<(AllocRange, SpanData)>,
+        matching_created: Option<(AllocRange, SpanData)>,
+        protected: Option<(SbTag, SpanData, SpanData)>,
+    },
+}
+
+pub trait GlobalStateExt {
+    fn current_span(&self, threads: &ThreadManager<'_, '_>) -> Span;
+
+    fn log_creation(
+        &mut self,
+        parent: Option<SbTag>,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn log_invalidation(
+        &mut self,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn log_protector(
+        &mut self,
+        orig_tag: SbTag,
+        tag: SbTag,
+        alloc: AllocId,
+        threads: &ThreadManager<'_, '_>,
+    );
+
+    fn get_stack_history(
+        &self,
+        tag: SbTag,
+        alloc: AllocId,
+        alloc_range: AllocRange,
+        offset: Size,
+        protector_tag: Option<SbTag>,
+    ) -> Option<TagHistory>;
+}
+
+impl GlobalStateExt for GlobalStateInner {
+    fn current_span(&self, threads: &ThreadManager<'_, '_>) -> Span {
+        threads
+            .active_thread_stack()
+            .into_iter()
+            .rev()
+            .find(|frame| {
+                let def_id = frame.instance.def_id();
+                def_id.is_local() || self.local_crates.contains(&def_id.krate)
+            })
+            .map(|frame| frame.current_span())
+            .unwrap_or(rustc_span::DUMMY_SP)
+    }
+
+    fn log_creation(
+        &mut self,
+        parent: Option<SbTag>,
+        tag: SbTag,
+        alloc: AllocId,
+        range: AllocRange,
+        threads: &ThreadManager<'_, '_>,
```

Comment:
I think this should just be a `Span` argument so the caller can provide it.

Reply:
The caller is in stacked_borrows.rs and I'd like to keep all `Span` logic out of that file if possible.

Reply:
makes sense... still weird to pass in the ThreadManager when what we want is a span. Maybe we can come up with some abstraction in the future. Fine for now then